04_mender_setup_env_grub_functions.cfg

```diff
@@ -0,0 +1,129 @@
+        fi
+        editing=0
+        save_env --file ${MENDER_LOCK2} editing
+    else
+        editing=invalid
+        load_env --skip-sig --file ${MENDER_LOCK1} editing
+        if [ "${editing}" != 0 ]; then
+            # See comment about "free form" variables near the top.
+            if [ "$check_signatures" = "enforce" ]; then
+                load_env --skip-sig --file ${MENDER_ENV2} bootcount mender_boot_part upgrade_available
+                save_env --file ${MENDER_ENV1} bootcount mender_boot_part upgrade_available
+            else
+                load_env --skip-sig --file ${MENDER_ENV2} bootcount mender_boot_part upgrade_available mender_systemd_machine_id
+                save_env --file ${MENDER_ENV1} bootcount mender_boot_part upgrade_available mender_systemd_machine_id
+            fi
+            editing=0
+            save_env --file ${MENDER_LOCK1} editing
+        fi
+    fi
+}
+
+function mender_save_env {
+    # Save redundant environment.
+    editing=1
+    save_env --file ${MENDER_LOCK2} editing
+    # See comment about "free form" variables near the top.
+    if [ "$check_signatures" = "enforce" ]; then
+        save_env --file ${MENDER_ENV2} bootcount mender_boot_part upgrade_available
+    else
+        save_env --file ${MENDER_ENV2} bootcount mender_boot_part upgrade_available mender_systemd_machine_id
+    fi
+    editing=0
+    save_env --file ${MENDER_LOCK2} editing
+
+    editing=1
+    save_env --file ${MENDER_LOCK1} editing
+    # See comment about "free form" variables near the top.
+    if [ "$check_signatures" = "enforce" ]; then
+        save_env --file ${MENDER_ENV1} bootcount mender_boot_part upgrade_available
+    else
+        save_env --file ${MENDER_ENV1} bootcount mender_boot_part upgrade_available mender_systemd_machine_id
+    fi
+    editing=0
+    save_env --file ${MENDER_LOCK1} editing
+}
+
+function mender_check_grubenv_valid {
+    if [ "${mender_boot_part}" != "${mender_rootfsa_part}" -a "${mender_boot_part}" != "${mender_rootfsb_part}" ]; then
+        return 1
+    fi
+
+    if [ "${bootcount}" != "0" -a "${bootcount}" != "1" ]; then
+        return 1
+    fi
+
+    if [ "${upgrade_available}" != "0" -a "${upgrade_available}" != "1" ]; then
+        return 1
+    fi
+
+    return 0
+}
+
+function mender_load_env {
+    # See comment about "free form" variables near the top.
+    if [ "$check_signatures" = "enforce" ]; then
+        load_env --skip-sig --file ${MENDER_ENV1} bootcount mender_boot_part upgrade_available
+    else
+        load_env --skip-sig --file ${MENDER_ENV1} bootcount mender_boot_part upgrade_available mender_systemd_machine_id
+    fi
+
+    if ! mender_check_grubenv_valid; then
+        if [ "${check_signatures}" == "enforce" ]; then
+            echo "Unverified environment and signatures enabled. Rebooting."
+            sleep 10
+            reboot
+        fi
+    fi
```

Comment:
Good point, there should be an error path here, actually. At this point there is little we can do, but I reckon that setting `mender_boot_part=${mender_rootfsa_part}` gives at least a chance that the system might boot, even if the partition may be wrong.

I don't think it can be used for downgrade attacks, because you'd need to have root access to corrupt the environment in this way, and if you had that you could just change the boot script directly.